### ISSUE TITLE
Major slowdown of count(*) after v12 to v16 upgrade

### ISSUE THREAD

**Issue by kmoppel** [laurenz/oracle_fdw#727]
Hi Laurenz! Stumbled on a weird issue for a client - tried all common tricks (chaning prefetch size, analyze, different Oracle instant client versions, re-creating the extension and foreign table, compiling oracle_fdw HEAD) but couldn't improve the situation - so any hints at what else to try, or how to debug it deeper would be very much appreciated.

![oracle_diag](https://github.com/user-attachments/assets/a6085ac5-0810-45e6-9f66-ed3b45a9a1fb)
QUERY: a super simple `select count(*) from tbl_with_116k_rows_on_oracle_side` 
BEFORE: PG v12.22, oracle_fdw 2.3.0devel, Oracle client 18.5 = sub-second runtime
AFTER: PG v16.6, oracle_fdw 2.7.0, latest Oracle client 23.6 = ~40s

Oracle version: 19.15.0.0
OS: Ubuntu 20.04

Explain plan was identical, a "SELECT /* ID */ '1' FROM TBL;" and hardly any activity visible on Oracle side.

**Comment by kmoppel** [comment]
Looks to be something network related as hinted by low (~15%) CPU usage and perf syscalls, strace output - see images below.  Also tried v16 +  old v18.5 Oracle client but still slow.

![Kuvatõmmis 2024-12-20 133133](https://github.com/user-attachments/assets/d6d84169-f361-4864-be51-d60441453fac)
![Kuvatõmmis 2024-12-20 133154](https://github.com/user-attachments/assets/271aff42-9a0b-44cd-a93d-c632bd253df1)
![strace](https://github.com/user-attachments/assets/9239b329-e89c-486f-b31e-44a456fe8181)

**Comment by laurenz** [comment]
Hm, there is a lot between 2.3.0 and 2.7.0, so it's hard to pinpoint.

One thing that would be interesting is to use the same Oracle versions for both tests. That would tell us if the problem is caused by changes on oracle_fdw or not.
And if you can reproduce the problem like that, you could use `git bisect` to pinpoint a certain oracle_fdw commit.

Is there a LOB column in the Oracle table? Can you reproduce the problem without a LOB column?
Did you set the `nchar` option of the foreign table?

**Comment by kajarkk** [comment]
I'm the one who have the issue. Just for extra info: no LOB columns and no nchar or any other options used for foreign server. Only option what is used is prefetch = '1000' for tables, but as kmoppel said prefetch doesnt change anything.

**Comment by kmoppel** [comment]
@laurenz Did some bisecting - and seems https://github.com/laurenz/oracle_fdw/commit/d679a73d02057fd674fc78fe5f25183e5c9a172b is to blame - causes a 2x slowdown on my good homelab network (Oracle and FDW on different nodes). Which I imagine could translate also into this larger observed difference on a slower network, given the kernel version (5.15.0-126) on the problematic systems is also considered very problematic according to google.

Quite some changes in that commit though so was not sadly able to pinpoint deeper on the fly.

**Comment by kmoppel** [comment]
Also an important thing I missed initially - the Oracle side table does contain a `SDO_GEOMETRY` field, disabling prefetch :/

And just in case - still no "pretty" looking fix possible to push down the common COUNT(*) scenario to Oracle? As mentioned here https://github.com/laurenz/oracle_fdw/issues/288#issuecomment-453461299

DESCRIBE Output, ~300 byte rows.
```
(
    MAJA_ID NUMBER(10,0),
    MAAKOND VARCHAR2(200),
    ASULA VARCHAR2(200),
    OMAVALITSUS VARCHAR2(200),
    TANAV_MAJA VARCHAR2(4000),
    EHR_KOOD VARCHAR2(100),
    KATASTRITUNNUS VARCHAR2(100),
    X NUMBER,
    Y NUMBER,
    KORRARV NUMBER(9,0),
    KORTARV NUMBER(9,0),
    TREPARV NUMBER(9,0),
    LEPING VARCHAR2(15),
    SEISUND VARCHAR2(15),
    PIIRKOND VARCHAR2(4000),
    KL_ARV NUMBER,
    KL_ARV_Z NUMBER,
    ADS_OID VARCHAR2(32),
    ALALIIK VARCHAR2(15),
    MA_ADR_ID NUMBER,
    ADOB_ID NUMBER,
    GEOMETRY MDSYS.SDO_GEOMETRY,
    ALGKPV DATE,
    TEHN_VOIMEKUS VARCHAR2(4000),
    LAIEND VARCHAR2(10),
    MUUDETUD_KP DATE,
    AD_OB_KEHTIVUS CHAR(7),
    STATUS VARCHAR2(10),
    TEHTKPV DATE,
    KAEV VARCHAR2(15),
    HF_SEISUND VARCHAR2(15),
    HF_KPV DATE,
    EP_SEISUND VARCHAR2(15),
    EP_KPV DATE
)
```

**Comment by laurenz** [comment]
Hm, that's big commit, yes.
And yes, `MDSYS.SDO_GEOMETRY` will disable prefetching. But that column is not used at all, so prefetching should still work.
Could you try a network analysis tool like Wireshark to see if there are any noticeable differences in the traffic? Perhaps there are more round trips, although I wouldn't understand why.

As a way forward, could you try with Git HEAD and set `prefetch` to a value higher than 1000? Perhaps that makes a difference, although I cannot imagine that it will make such a big difference.

Other than that, I am somewhat at loss.

**Comment by kajarkk** [comment]
I found that geometry column was the troublemaker. For a test I created foreign table without geometry and used also OPTIONS(table '(SELECT field1,field2... FROM schema.table) and performance seems to be same as it was Postgresql 12 version.

**Comment by kmoppel** [comment]
And with HEAD - setting `prefetch` over 1000 indeed helped only minimally for the original SELECT count(*)

**Comment by laurenz** [comment]
Strange. I have no explanation for this performance regression, since I did not modify the code to access geometries.
I don't know how to help you with that... perhaps you can run a network trace to see if the data exchanged are different.

**Comment by laurenz** [comment]
Did you get to investigate the case with Wireshark or some other network tracer?

**Comment by kajarkk** [comment]
> Did you get to investigate the case with Wireshark or some other network tracer?

No, I went the another way and skipped using oracle_fdw for now,I use other method to bring data from Oracle to Postgresql.

**Comment by laurenz** [comment]
Then I'll close the issue, ok?
Sorry that we could not get to the bottom of this.

**Comment by kajarkk** [comment]
Yes I think so.
